src/Client/Invoice.php

```
@@ -6,38 +6,85 @@
+     * @param string                                           $storeId
+     * @param \BTCPayServer\Util\PreciseNumber                 $amount
+     * @param string                                           $currency
+     * @param string|null                                      $orderId
+     * @param string|null                                      $customerEmail
+     * @param array|null                                       $metaData
+     * @param \BTCPayServer\Client\InvoiceCheckoutOptions|null $checkoutOptions
+     *
+     * @return \BTCPayServer\Result\Invoice
+     * @throws \JsonException
+     */
+    public function createInvoice(
+        string $storeId,
+        PreciseNumber $amount,
+        string $currency,
+        ?string $orderId = '',
+        ?string $customerEmail = '',
+        ?array $metaData = [],
+        ?InvoiceCheckoutOptions $checkoutOptions = null
+    ): \BTCPayServer\Result\Invoice {
         // TODO test & finish this
         // TODO add parameter for metadata and merge with orderId and buyerEmail
-        $url = $this->getBaseUrl() . 'stores/' . urlencode($storeId) . '/invoices';
+        $url = $this->getBaseUrl() . 'stores/' . urlencode(
+            $storeId
+        ) . '/invoices';
         $headers = $this->getRequestHeaders();
         $method = 'POST';
 
-        $body = json_encode([
-            'amount' => $amount,
+        // Prepare metadata.
+        // Todo: decide if explicit param or metadata array takes precedence for
+        // email and order id doing array_merge or similar. Current rule explicit
+        // params take precedence.
+        $metaDataMerged = [];
+
+        // Set metaData if any.
+        if ($metaData) {
+            $metaDataMerged = $metaData;
+        }
+
+        // We need to check $orderId and $customerEmail explicitly as they are optional.
+        // Overwrites existing data passed by $metaData.
+        if ($orderId) {
+            $metaDataMerged['orderId'] = $orderId;
```

Should we throw an exception if a previous value was already there? You may be overwriting something without the dev realizing...

---

Ok a good idea I think, so we need a check for those two metadata keys and throw an exception.

---

I'd use `array_key_exists` for this, not `isset()` :wink: